crates/prune/prune/src/segments/storage_history.rs

```diff
@@ -67,34 +69,52 @@ impl<DB: Database> Segment<DB> for StorageHistory {
         }
 
         let mut last_changeset_pruned_block = None;
+        // Deleted storage changeset keys (account addresses and storage slots) with the highest
+        // block number deleted for that key
+        let mut highest_deleted_storages = FxHashMap::default();
         let (pruned_changesets, done) = provider
             .prune_table_with_range::<tables::StorageChangeSets>(
                 BlockNumberAddress::range(range),
                 &mut limiter,
                 |_| false,
-                |row| last_changeset_pruned_block = Some(row.0.block_number()),
+                |(BlockNumberAddress((block_number, address)), entry)| {
+                    highest_deleted_storages.insert((address, entry.key), block_number);
+                    last_changeset_pruned_block = Some(block_number);
+                },
             )?;
         trace!(target: "pruner", deleted = %pruned_changesets, %done, "Pruned storage history (changesets)");
 
         let last_changeset_pruned_block = last_changeset_pruned_block
-            // If there's more storage storage changesets to prune, set the checkpoint block number
-            // to previous, so we could finish pruning its storage changesets on the next run.
+            // If there's more storage changesets to prune, set the checkpoint block number to
+            // previous, so we could finish pruning its storage changesets on the next run.
             .map(|block_number| if done { block_number } else { block_number.saturating_sub(1) })
             .unwrap_or(range_end);
 
-        let (processed, pruned_indices) = prune_history_indices::<DB, tables::StoragesHistory, _>(
+        // Sort highest deleted block numbers by account address and storage key and turn them into
+        // sharded keys.
+        // We did not use `BTreeMap` from the beginning, because it's inefficient for hashes.
+        let highest_sharded_keys = highest_deleted_storages
+            .into_iter()
+            .sorted_unstable() // Unstable is fine because no equal keys exist in the map
```

Comment:
think no need for parallel here

Reply:
why not max perf from start? is there much overhead in using the par sort?

Reply:
it can actually make the perf worse, the unit of work should be big enough to get a benefit from parallelization. i believe here it's not that big, given the size of the array and the simple nature of sorting